tests/shared/src/test/scala/scala/meta/tests/contrib/AssociatedCommentsTest.scala

```diff
@@ -39,19 +38,234 @@ class AssociatedCommentsTest extends FunSuite {
+
+    val defnClass = input.find(_.is[Defn.Class]).get
+    val defnObject = input.find(_.is[Defn.Object]).get
+
+    assertExpectations(input)(
+      leading = Map(
+        defnClass -> Set(
+          """|/** Scaladoc for class A
+             |  */""".stripMargin
+        ),
+        defnObject -> Set(
+          """|/** Scaladoc for object A
+             |  */""".stripMargin
+        )
+      )
+    )
+  }
+
+  test("single leading comment at the beginning of a file") {
+    val input =
+      """|// leading
+         |object A
+         |""".stripMargin.parse[Source].get
+
+    val defnObject = input.find(_.is[Defn.Object]).get
+
+    assertExpectations(input)(
+      leading = Map(
+        defnObject -> Set("// leading")
+      )
+    )
+  }
+
+  test("multiple leading comments in a single line at the beginning of a file") {
+    val input =
+      """|/** leading 1 */ /* leading 2 */ // leading 3
+         |class A
+         |""".stripMargin.parse[Source].get
+
+    val defnClass = input.find(_.is[Defn.Class]).get
+
+    assertExpectations(input)(
+      leading = Map(
+        defnClass -> Set("/** leading 1 */", "/* leading 2 */", "// leading 3")
+      )
+    )
+  }
+
+  ignore("multiple leading comments in different lines at the beginning of a file") {
+    val input =
+      """|/** leading 1 */
+         |/* leading 2 */
+         |// leading 3
+         |trait A
+         |""".stripMargin.parse[Source].get
+
+    val defnTrait = input.find(_.is[Defn.Trait]).get
+
+    assertExpectations(input)(
+      leading = Map(
+        defnTrait -> Set("/** leading 1 */", "/* leading 2 */", "// leading 3")
+      )
+    )
+  }
+
+  ignore("single trailing comment at the end of a file") {
+    val input =
+      """|object A {
+         |} // trailing
+         |""".stripMargin.parse[Source].get
+
+    val defnObject = input.find(_.is[Defn.Object]).get
+    val template = input.find(_.is[Template]).get // overlaps with `object`
+
+    assertExpectations(input)(
+      trailing = Map(
+        defnObject -> Set("// trailing"),
+        template -> Set("// trailing")
+      )
+    )
+  }
+
+  test("multiple trailing comments in a single line at the end of a file") {
+    val input =
+      """|class A {
+         |} /** trailing 1 */ /* trailing 2 */ // trailing 3
+         |""".stripMargin.parse[Source].get
+
+    val defnClass = input.find(_.is[Defn.Class]).get
+    val template = input.find(_.is[Template]).get // overlaps with `class`
+
+    val expectedComments = Set("/** trailing 1 */", "/* trailing 2 */", "// trailing 3")
+    assertExpectations(input)(
+      trailing = Map(
+        defnClass -> expectedComments,
+        template -> expectedComments
+      )
+    )
+  }
+
+  test("multiple trailing comments in different lines at the end of a file") {
+    val input =
+      """|trait A {
+         |} /** trailing 1 */
+         |/* trailing 2 */
+         |// trailing 3
+         |""".stripMargin.parse[Source].get
```

Comment:
the fix makes this scenario consistent with the one where multiple leading comments appear after BOF. This means that all comments (regardless of how many lines) are associated to the closest non-trivia token. Please let me know if you think it is not correct

Reply:
This looks great!